xen/arch/arm/mm.c

```diff
@@ -1287,24 +1287,55 @@ long arch_memory_op(int op, XEN_GUEST_HANDLE_PARAM(void) arg)
         if ( d == NULL )
             return -ESRCH;
 
-        for ( i = 0; i < req.num_frames; i++ )
+        if ( req.num_frames == 1 )
         {
+            /* more than 80% cases for GSX operation */
             xen_pfn_t pa, ma;
 
-            if ( unlikely(copy_from_guest_offset(&pa, req.pa, i, 1)) )
+            if ( unlikely(copy_from_guest(&pa, req.pa, 1)) )
             {
                 rcu_unlock_domain(d);
                 return -EFAULT;
             }
 
             ma = pfn_to_paddr(p2m_lookup(d, paddr_to_pfn(pa), NULL));
 
-            if ( unlikely(copy_to_guest_offset(req.ma, i, &ma, 1)) )
+            if ( unlikely(copy_to_guest(req.ma, &ma, 1)) )
             {
                 rcu_unlock_domain(d);
                 return -EFAULT;
             }
         }
+        else
+        {
+            xen_pfn_t *pma;
+
+            pma = xmalloc_array(xen_pfn_t, req.num_frames);
```

Comment:
Wouldn't be more correct to use alloc_domheap_pages here? I am afraid you may exhaust Xen heap using xmalloc for such a big allocation.

Reply:
for small allocations we can even alloc on stack, e.g. for number of frames less than 16/32 just use a local array

Reply:
Yes, sure.

Reply:
The maximum allocation I have faced this far is 16 pages, is it that big? Also there is an example where xmalloc_x is used during domain hypercalls servicing.

Reply:
64 KB? Perhaps not so big, as I imagined myself the first time.

Reply:
It was one time of 8K frames (gfn's)  to lookup for heap.

Reply:
~2K frames are requested to translate on application switch in DomA (full screen redraw). With current resolutions, for sure.